File: packages/core/src/__tests__/engine.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { Engine, Skip } from '../engine.js'

describe('Engine', () => {
  let engine: Engine

  beforeEach(() => {
    engine = new Engine()
  })

  it('should create event types with unique names', () => {
    const a = engine.event('click')
    const b = engine.event('click')
    expect(a.name).not.toBe(b.name)
    expect(a.name).toContain('click')
  })

  it('should emit and receive events via on()', () => {
    const ev = engine.event<number>('num')
    const results: number[] = []

    engine.on(ev, (n) => results.push(n))

    engine.emit(ev, 42)
    expect(results).toEqual([42])
  })

  it('should support on() with unsubscribe', () => {
    const ev = engine.event<string>('msg')
    const results: string[] = []

    const unsub = engine.on(ev, (msg) => results.push(msg))

    engine.emit(ev, 'hello')
    expect(results).toEqual(['hello'])

    unsub()

    engine.emit(ev, 'world')
    expect(results).toEqual(['hello']) // no change
  })

  it('should support multiple handlers on the same event type', () => {
    const ev = engine.event<number>('num')
    const r1: number[] = []
    const r2: number[] = []

    engine.on(ev, (n) => r1.push(n))
    engine.on(ev, (n) => r2.push(n * 2))

    engine.emit(ev, 3)
    expect(r1).toEqual([3])
    expect(r2).toEqual([6])
  })

  it('should handle events with no consumers gracefully', () => {
    const orphan = engine.event<number>('orphan')
    expect(() => engine.emit(orphan, 99)).not.toThrow()
  })

  it('should handle nested emit during propagation', () => {
    const a = engine.event<number>('a')
    const results: number[] = []

    engine.on(a, (val) => {
      results.push(val)
      if (val < 3) {
        engine.emit(a, val + 1)
      }
    })

    engine.emit(a, 1)
    expect(results).toContain(1)
    expect(results).toContain(2)
    expect(results).toContain(3)
  })

  it('should chain events via emit inside on()', () => {
    const a = engine.event<number>('a')
    const b = engine.event<number>('b')
    const c = engine.event<number>('c')
    const results: number[] = []

    engine.on(a, (x) => {
      engine.emit(b, x + 1)
    })
    engine.on(b, (x) => {
      engine.emit(c, x * 10)
    })
    engine.on(c, (val) => results.push(val))

    engine.emit(a, 1)
    expect(results).toEqual([20]) // (1+1)*10
  })

  it('should handle fan-out (one event, two on() handlers emitting to different events)', () => {
    const src = engine.event<number>('src')
    const out1 = engine.event<number>('out1')
    const out2 = engine.event<number>('out2')
    const r1: number[] = []
    const r2: number[] = []

    engine.on(src, (x) => engine.emit(out1, x + 10))
    engine.on(src, (x) => engine.emit(out2, x + 20))
    engine.on(out1, (v) => r1.push(v))
    engine.on(out2, (v) => r2.push(v))

    engine.emit(src, 1)
    expect(r1).toEqual([11])
    expect(r2).toEqual([21])
  })

  it('should expose getRules()', () => {
    const a = engine.event('a')
    engine.on(a, () => {})
    expect(engine.getRules().length).toBeGreaterThanOrEqual(1)
  })

  it('should expose getMailboxes()', () => {
    const a = engine.event<number>('a')
    engine.emit(a, 1)
    expect(engine.getMailboxes()).toBeDefined()
  })

  it('should expose getDAG()', () => {
    const a = engine.event('a')
    engine.on(a, () => {})
    const dag = engine.getDAG()
    expect(dag.nodes.length).toBeGreaterThanOrEqual(1)
  })

  it('should handle multiple sequential emits independently', () => {
    const ev = engine.event<number>('ev')
    const results: number[] = []

    engine.on(ev, (v) => results.push(v))

    engine.emit(ev, 1)
    engine.emit(ev, 2)
    engine.emit(ev, 3)

    expect(results).toEqual([1, 2, 3])
  })

  it('should export Skip sentinel', () => {
    expect(Skip).toBe(Symbol.for('pulse.skip'))
  })

  it('should destroy and stop processing', () => {
    const ev = engine.event<number>('ev')
    const results: number[] = []
    engine.on(ev, (v) => results.push(v))

    engine.emit(ev, 1)
    expect(results).toEqual([1])

    engine.destroy()

    // After destroy, emit should not call handler (no consumers registered)
    engine.emit(ev, 2)
    expect(results).toEqual([1])
  })

  it('should support on([A, B]) join syntax', () => {
    const a = engine.event<number>('a')
    const b = engine.event<string>('b')
    const results: string[] = []

    engine.on([a, b], (numVal: number, strVal: string) => {
      results.push(`${numVal}-${strVal}`)
    })

    // Only emit a — join should not fire
    engine.emit(a, 42)
    expect(results).toEqual([])

    // Now emit b — join should fire
    engine.emit(b, 'hello')
    expect(results).toEqual(['42-hello'])
  })

  it('should not leak events when emitting to event with no consumers', () => {
    const orphan = engine.event<number>('orphan')
    // No on() registered for orphan — zero consumers
    // This should not leak pool slots
    for (let i = 0; i < 100; i++) {
      engine.emit(orphan, i)
    }
    // If pool leak existed, 100 events would be lost from the pool
    // No assertion needed — this verifies no throw and no unbounded growth
    // The mailbox for orphan should be empty (no consumers = no enqueue)
    const mb = engine.getMailboxes().get(orphan)
    expect(mb).toBeUndefined() // never even created a mailbox
  })
})

File: packages/dart/test/engine_test.dart
import 'package:test/test.dart';
import 'package:pulse/pulse.dart';

void main() {
  late Engine engine;

  setUp(() {
    engine = Engine();
  });

  test('should create event types with unique names', () {
    final a = engine.event('click');
    final b = engine.event('click');
    expect(a.name, isNot(equals(b.name)));
    expect(a.name, contains('click'));
  });

  test('should pipe events from input to output', () {
    final input = engine.event<int>('input');
    final output = engine.event<int>('output');
    final results = <int>[];

    engine.pipe<int, int>(input, output, (x) => x * 2);
    engine.on<int>(output, (val) => results.add(val));

    engine.emit(input, 5);
    expect(results, equals([10]));
  });

  test('should chain multiple pipes', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final c = engine.event<int>('c');
    final results = <int>[];

    engine.pipe<int, int>(a, b, (x) => x + 1);
    engine.pipe<int, int>(b, c, (x) => x * 10);
    engine.on<int>(c, (val) => results.add(val));

    engine.emit(a, 1);
    expect(results, equals([20])); // (1+1)*10
  });

  test('should handle pipe returning null (no output emitted)', () {
    final input = engine.event<int>('input');
    final output = engine.event<int>('output');
    final results = <int>[];

    engine.pipe<int, int?>(input, output, (x) {
      if (x > 0) return x;
      return null;
    });
    engine.on<int>(output, (val) => results.add(val));

    engine.emit(input, -1);
    expect(results, equals([]));

    engine.emit(input, 5);
    expect(results, equals([5]));
  });

  test('should support on() with unsubscribe', () {
    final ev = engine.event<String>('msg');
    final results = <String>[];

    final unsub = engine.on<String>(ev, (msg) => results.add(msg));

    engine.emit(ev, 'hello');
    expect(results, equals(['hello']));

    unsub();

    engine.emit(ev, 'world');
    expect(results, equals(['hello'])); // no change
  });

  test('should support multiple handlers on the same event type', () {
    final ev = engine.event<int>('num');
    final r1 = <int>[];
    final r2 = <int>[];

    engine.on<int>(ev, (n) => r1.add(n));
    engine.on<int>(ev, (n) => r2.add(n * 2));

    engine.emit(ev, 3);
    expect(r1, equals([3]));
    expect(r2, equals([6]));
  });

  test('should resolve entire graph synchronously in a single emit', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final c = engine.event<int>('c');
    final d = engine.event<int>('d');
    final trace = <String>[];

    engine.pipe<int, int>(a, b, (x) {
      trace.add('a->b');
      return x + 1;
    });
    engine.pipe<int, int>(b, c, (x) {
      trace.add('b->c');
      return x + 1;
    });
    engine.pipe<int, int>(b, d, (x) {
      trace.add('b->d');
      return x * 2;
    });
    engine.on<int>(c, (_) => trace.add('c'));
    engine.on<int>(d, (_) => trace.add('d'));

    engine.emit(a, 0);
    expect(trace, contains('a->b'));
    expect(trace, contains('b->c'));
    expect(trace, contains('b->d'));
    expect(trace, contains('c'));
    expect(trace, contains('d'));
  });

  test('should handle nested emit during propagation', () {
    final a = engine.event<int>('a');
    final results = <int>[];

    engine.on<int>(a, (val) {
      results.add(val);
      if (val < 3) {
        engine.emit(a, val + 1);
      }
    });

    engine.emit(a, 1);
    expect(results, contains(1));
    expect(results, contains(2));
    expect(results, contains(3));
  });

  test('should handle pipe to multiple outputs', () {
    final input = engine.event<int>('input');
    final outA = engine.event<int>('outA');
    final outB = engine.event<int>('outB');
    final resA = <int>[];
    final resB = <int>[];

    engine.pipe<int, List<int>>(input, [outA, outB], (x) => [x + 1, x + 2]);
    engine.on<int>(outA, (v) => resA.add(v));
    engine.on<int>(outB, (v) => resB.add(v));

    engine.emit(input, 10);
    expect(resA, equals([11]));
    expect(resB, equals([12]));
  });

  test('should expose getRules()', () {
    final a = engine.event('a');
    final b = engine.event('b');
    engine.pipe(a, b, (x) => x);
    expect(engine.getRules().length, greaterThanOrEqualTo(1));
  });

  test('should expose getMailboxes()', () {
    final a = engine.event<int>('a');
    engine.emit(a, 1);
    expect(engine.getMailboxes(), isNotNull);
  });

  test('should expose getDAG()', () {
    final a = engine.event('a');
    final b = engine.event('b');
    engine.pipe(a, b, (x) => x);
    final dag = engine.getDAG();
    expect(dag.nodes.length, greaterThanOrEqualTo(1));
  });

  test('should handle events with no consumers gracefully', () {
    final orphan = engine.event<int>('orphan');
    expect(() => engine.emit(orphan, 99), returnsNormally);
  });

  test('should handle pipe unsubscribe', () {
    final a = engine.event<int>('a');
    final b = engine.event<int>('b');
    final results = <int>[];

    final unsub = engine.pipe<int, int>(a, b, (x) => x * 2);
    engine.on<int>(b, (v) => results.add(v));

    engine.emit(a, 5);
    expect(results, equals([10]));

    unsub();
    engine.emit(a, 5);
    expect(results, equals([10]));
  });

  test('should destroy and clean up', () {
    final a = engine.event<int>('a');
    final results = <int>[];
    engine.on<int>(a, (v) => results.add(v));

    engine.emit(a, 1);
    expect(results, equals([1]));

    engine.destroy();

    // After destroy, emitting should not reach the handler
    // (rules are removed, so no consumers).
    engine = Engine();
    final a2 = engine.event<int>('a2');
    engine.emit(a2, 2);
    expect(results, equals([1]));
  });

  test('should support Skip sentinel', () {
    final input = engine.event<int>('input');
    final output = engine.event<int>('output');
    final results = <int>[];

    engine.pipe<int, dynamic>(input, output, (x) {
      if (x < 0) return Skip;
      return x;
    });
    engine.on<int>(output, (v) => results.add(v));

    engine.emit(input, -1);
    expect(results, isEmpty);

    engine.emit(input, 5);
    expect(results, equals([5]));
  });
}
